### PR TITLE
[CA-1541] Address PPW Feature Branch Feedback

### PR DESCRIPTION
**Pull request by ehigham**
RR: https://broadworkbench.atlassian.net/browse/CA-1541
In this PR I'm addressing some of @zarsky-broad's suggestions and feedback in #2612, namely merging the Project per Workspace (PPW) branch, `v2_workspaces` into `dev`.
Isaac and I spoke at length about #2612 and a few suggestions for tidy-ups were also raised in `ajax.js`. I've tried to address these as well.

### CODE REVIEW

File: src/pages/billing/Project.js [hunk 1]
```diff
@@ -147,41 +147,39 @@ const BillingAccountSummaryPanel = (() => {
...
 
 const groupByBillingAccountStatus = (billingProject, workspaces) => {
-  const group = workspace => Utils.cond(
-    [billingProject.billingAccount === workspace.billingAccount, () => 'done'],
-    ['billingAccountErrorMessage' in workspace, () => 'error'],
-    [Utils.DEFAULT, () => 'updating']
-  )
+  const group = workspace => billingProject.billingAccount === workspace.billingAccount ? 'done' :
+    !!workspace.billingAccountErrorMessage ? 'error' :
+      'updating'
```
Comment (ehigham) [inline src/pages/billing/Project.js]: I went with ternary instead of `Utils.cond` because this is more idomatic for the language. I believe that newcomers who don't know Clojure would be far more familiar with C-like ternaries.

File: src/pages/billing/Project.js [hunk 2]
```diff
@@ -147,41 +147,39 @@ const BillingAccountSummaryPanel = (() => {
...
 
+  // return `Set`s of workspaces mapped by their status to reduce the complexity of looking up a
+  // workspace's status.
```
Comment (ehigham) [inline src/pages/billing/Project.js]: This comment is terrible. What I'm trying to get at is I'm using this data structure for two things - to test if we need to draw icons to represent the status of billing account changes and if we do, what icon to draw for each workspace. I'm using sets to reduce the complexity of seaching for the status of any workspace from `O(N * W)` to `O(N * 1)`, where `N` is the number of statuses a billing account change could have and `W` is the number of workspaces in a billing project. Note I need to perform this `W` times for each billing project and `W` can be very large in the case of GP's billing projects. Thus reduces the complexity by an order of magnitute.

Comment (ehigham) [reply src/pages/billing/Project.js]: (Suggestions welcome)

File: src/libs/ajax.js [hunk 3]
```diff
@@ -471,43 +480,34 @@ const Billing = signal => ({
...
 
-  billingProject: async projectName => {
-    const route = `billing/v2/${projectName}`
-    const res = await fetchRawls(route, _.merge(authOpts(), { signal, method: 'GET' }))
+  listProjectUsers: async projectName => {
+    const res = await fetchRawls(`billing/v2/${projectName}/members`, _.merge(authOpts(), { signal }))
     return res.json()
   },
 
-  project: projectName => {
```
Comment (s-rubenstein) [inline src/libs/ajax.js:480]: Why are we getting rid of the project wrapper here? That seemed more representative of the codebase's general practice. (Similar to analysis and notebooks under storage)

File: src/pages/billing/Project.js [hunk 4]
```diff
@@ -60,7 +60,7 @@ const WorkspaceCardHeaders = Utils.memoWithName('WorkspaceCardHeaders', ({ needs
   ])
 })
 
-const ExpandedInfoRow = Utils.memoWithName('ExpandedInfoRow', ({ title, details, errorMessage }) => {
```
Comment (s-rubenstein) [inline src/pages/billing/Project.js:63]: Can we confirm that without this it shows up with the right name in the React dev tools?

File: src/pages/billing/Project.js (hunk 2 above)
Comment (s-rubenstein) [reply src/pages/billing/Project.js]: Could you say that last bit that you had there? 
```
I'm using sets to reduce the complexity of seaching for the status of any workspace from O(N * W) to O(N * 1), where N is the number of statuses a billing account change could have and W is the number of workspaces in a billing project. Note I need to perform this W times for each billing project and W can be very large in the case of GP's billing projects. Thus reduces the complexity by an order of magnitute.
```

File: src/libs/ajax.js (hunk 3 above)
Comment (ehigham) [reply src/libs/ajax.js:480]: I did do that originally (renaming it to `projectUsers`) but then thought the wrapper served very little benefit. Merging with the `Billing` object and renaming seemed cleaner.

File: src/pages/billing/Project.js (hunk 4 above)
Comment (ehigham) [reply src/pages/billing/Project.js:63]: Sure!

Comment (ehigham) [reply src/pages/billing/Project.js:63]: Confirmed - it's called the right thing.

File: src/pages/billing/Project.js (hunk 2 above)
Comment (ehigham) [reply src/pages/billing/Project.js]: Thanks!

**Review by marctalbott — APPROVED** [review]

File: src/components/NewWorkspaceModal.js [hunk 5]
```diff
@@ -161,18 +159,18 @@ const NewWorkspaceModal = Utils.withDisplayName('NewWorkspaceModal', ({
...
           onChange: ({ value }) => setNamespace(value),
           styles: { option: provided => ({ ...provided, padding: 0 }) },
-          options: _.uniq(_.map(({ projectName, invalidBillingAccount }) => ({
-            label: h(Fragment, [
-              invalidBillingAccount ?
+          options: _.map(({ projectName, invalidBillingAccount }) => ({
+            label: ((label = div({ style: { padding: 10 } }, [projectName])) => {
+              return invalidBillingAccount ?
                 h(TooltipTrigger, {
                   content: 'Workspaces may only be created in billing projects that have a Google billing account accessible in Terra',
```
Comment (zarsky-broad) [inline src/components/NewWorkspaceModal.js]: You can just make content `invalidBillingAccount && 'Workspaces...'` here, and remove the ternary altogether; a tooltip won't be rendered if content is falsy. Why did you make label an argument to this function if it's never called with arguments? I continue to not like the self-invoking function pattern.

File: src/libs/ajax.js [hunk 6]
```diff
@@ -442,15 +442,24 @@ const Billing = signal => ({
...
 
   createProject: async (projectName, billingAccount) => {
-    const res = await fetchRawls('billing/v2',
-      _.mergeAll([authOpts(), jsonBody({ projectName, billingAccount }), { signal, method: 'POST' }]))
-    return res
+    const res = await fetchRawls('billing/v2', _.mergeAll([
+      authOpts(),
+      jsonBody({ projectName, billingAccount }),
+      { signal, method: 'POST' }
+    ]))
```
Comment (zarsky-broad) [inline src/libs/ajax.js]: I don't love breaking this into 3 lines; I cold see this all on one line, but I think if we're adding linebreaks I might go for 
```js
fetchRawls(
  'path',
  _.mergeAll([])
)
```

File: src/libs/ajax.js (hunk 3 above)
Comment (zarsky-broad) [reply src/libs/ajax.js:480]: I'm with @s-rubenstein on this one, I find the consistency with other places to be extremely valuable. It also cuts down on the number of positional arguments to functions.

File: src/libs/ajax.js [hunk 7]
```diff
@@ -471,43 +480,34 @@ const Billing = signal => ({
...
+  },
 
-      changeUserRoles: async (email, oldRoles, newRoles) => {
-        if (!_.isEqual(oldRoles, newRoles)) {
-          await Promise.all(_.map(role => addRole(role, email), _.difference(newRoles, oldRoles)))
-          return Promise.all(_.map(role => removeRole(role, email), _.difference(oldRoles, newRoles)))
-        }
-      }
+  changeUserRoles: async (projectName, email, oldRoles, newRoles) => {
+    const billing = Billing(signal)
```
Comment (zarsky-broad) [inline src/libs/ajax.js]: This is a potentially destructive operation, so it shouldn't take `signal`, it should complete even if a modal or something is closed.

File: src/pages/billing/Project.js [hunk 8]
```diff
@@ -29,14 +29,14 @@ const workspaceLastModifiedWidth = 150
 const workspaceExpandIconSize = 20
 const billingAccountIconSize = 16
 
-const BillingAccountIcon = {
+const billingAccountIcon = {
```
Comment (zarsky-broad) [inline src/pages/billing/Project.js]: Would you consider `billingAccountIcons`, since it's effectively a dict?

File: src/pages/billing/Project.js (hunk 1 above)
Comment (zarsky-broad) [reply src/pages/billing/Project.js]: I hear you, but we really try to avoid nested ternaries. I'd prefer to see it as a cond because that's our standard in this codebase, and we use it everywhere else. I want to keep things consistent.

File: src/pages/billing/Project.js [hunk 9]
```diff
@@ -320,15 +322,15 @@ const ProjectDetail = ({ billingProject, reloadBillingProject, billingAccounts,
...
+  )(() => Ajax(signal).Billing.listProjectUsers(billingProject.projectName)
     .then(collectUserRoles)
     .then(setProjectUsers)
   )
 
   const removeUserFromBillingProject = _.flow(
     reportErrorAndRethrow('Error removing member from billing project'),
     Utils.withBusyState(setUpdating)
-  )(Ajax().Billing.project(billingProject.projectName).removeUser)
+  )(_.partial(Ajax().Billing.removeProjectUser, [billingProject.projectName]))
```
Comment (zarsky-broad) [inline src/pages/billing/Project.js:335]: So these are our first instances of `_.partial` in the codebase; we normally declare any function likely to be used like this with `_.curry` instead. That said, this feels to me like a reason to revert the structure in the ajax file to how it was before, which was done with the intention of how it was going to be used in mind.

**Review by zarsky-broad — COMMENTED** [review]
I'm off on vacation after today, so I'm gonna nominate @petesantos to follow up. Here are my thoughts, though; thank you for sticking with this!

File: src/pages/billing/Project.js (hunk 1 above)
Comment (ehigham) [reply src/pages/billing/Project.js]: Cond is noisy. Cond does not preserve the evaluation order of ternaries. Ternaries are what programmers familiar with C syntax are familiar with. We should be consistent with the idoms of the language and stop pretending this is clojure.

File: src/libs/ajax.js (hunk 3 above)
Comment (ehigham) [reply src/libs/ajax.js:480]: I disagree strongly. This way keeps things simple and follows the underlying API whereas the other feels consistency for consistency's sake. I don't really know/understand what this is done for either.

File: src/components/NewWorkspaceModal.js (hunk 5 above)
Comment (ehigham) [reply src/components/NewWorkspaceModal.js]: I didn't like `div({ style: { padding: 10 } }, [projectName])` appearing twice. What I really wanted to do was use a let expression to generate a binding right next to where it's used. I didn't want to make a whole bunch of changes here like restructuring the code and adding a couple of `const`s a thousand lines above this as I prefer to keep helper definitions close to uses. Sadly, JavaScript lacks `let` expressions but it does have lambdas and `let`s are syntactic sugar for lambdas. Thus, I can encode
```
let x = <init> in
   f(x, x)
```
simply by
```javascript
((x = <init>) => f(x, x))()
```
I'm sorry you don't like self-invoking functions - they're such a useful tool.

Comment (ehigham) [reply src/components/NewWorkspaceModal.js]: Thanks for the tip about tooltips not rendering if the content is falsy. I didn't realise that was a thing.

File: src/libs/ajax.js (hunk 6 above)
Comment (ehigham) [reply src/libs/ajax.js]: I don't love lines being 150 characters long. I don't particularly like lines exceeding 100 columns either. I'm just going to revert this.

File: src/pages/billing/Project.js (hunk 9 above)
Comment (ehigham) [reply src/pages/billing/Project.js:335]: If using the `project` wrapper was a means of achieving partial application then that's another reason against it. As a side note, I don't like this practice of defining curried functions. If it's a normal function then I can choose what to do with it. If I want to, I can curry it or partially apply arguments or just call it without a ton of parentheses. Let the caller decide.

File: src/libs/ajax.js (hunk 7 above)
Comment (ehigham) [reply src/libs/ajax.js]: Can you explain what this means? Why is this destructive? I'm not even sure what the `signal` is for...

File: src/pages/billing/Project.js (hunk 9 above)
Comment (s-rubenstein) [reply src/pages/billing/Project.js:335]: So, having now looked at the docs, I understand what this is doing but I found it less readable, personally, then the previous code. Especially in the case of the props below. However, in this case I _personally_ don't have strong opinions on it, but just wanted to throw my two cents in.

File: src/pages/billing/Project.js (hunk 1 above)
Comment (s-rubenstein) [reply src/pages/billing/Project.js]: General question: why should we be preferring C-style syntax over clojure-style syntax?

I am happy with the argument in this case that it preserves evaluation order, that is compelling to me personally, but I don't necessarily find C that readable and the nice thing about the clojure syntax is it makes it a little bit more clear to me what the possible outcomes are, whereas if we got more nested ternaries in here (two is okay, but I have seen it get much longer than that), I personally start to have trouble making that out. It takes what was a relatively easy task and makes it difficult.

File: src/libs/ajax.js (hunk 3 above)
Comment (s-rubenstein) [reply src/libs/ajax.js:480]: The benefit that it provides, in my mind, is similar to object layering. By abstracting out `projectUsers` (and I like that name a lot), we gain the ability to have a structure that groups more in a way that makes it easy to find what neighboring functions are, and keeps associated endpoints grouped. (If I see a `.projectUsers(project).add()` call, I can fairly easily follow `.projectUsers` to find all associated endpoints.

I would personally prefer those child objects pulled out into their own objects, so we get a cleaner view in our editors, which I think would more easily show the benefits here.

I think it does follow the underlying REST structure a bit more (where they are all a page) but it doesn't necessarily follow the route definitions in the service as cleanly. In Rawls (or Sam) these are grouped effectively by arg, so the original code here preserves that, IMO.

Comment (s-rubenstein) [reply src/libs/ajax.js:480]: Thinking more about it, the difference between `projectUsers` and `notebooks`/`analysis` is that notebooks and analysis provide useful wrappers around some business logic (that they are making storage api calls and doing a bit of work around abstracting that), while `projectUsers` provides an abstraction that doesn't help much and adds potential for misunderstanding.

`workspaces` is a place which might be more comparable, and is an interesting case study, but that has multiple params and specific structure to it, whereas this just has one route param.

**Review by s-rubenstein — APPROVED** [review]
Pending the curry vs partial decision, I am happy with this as is.

File: src/pages/billing/Project.js (hunk 1 above)
Comment (ehigham) [reply src/pages/billing/Project.js]: Parking this conversation for a UI meeting. Lots of good comments.Reverting to cond.

File: src/libs/ajax.js (hunk 7 above)
Comment (ehigham) [reply src/libs/ajax.js]: @petesantos explained - thanks very much.

File: src/pages/billing/Project.js (hunk 9 above)
Comment (ehigham) [reply src/pages/billing/Project.js:335]: Follow-up - I didn't realise lodash's `curry` allowed you to pass more than one argument at a time (ie `f(a)(b)(c)` can be called `f(a, b, c)` when `f` is `_.curry`'ed.

File: src/libs/ajax.js (hunk 3 above)
Comment (petesantos) [reply src/libs/ajax.js:480]: In this particular case, I do think it comes down to how it is being used. It does seem odd that when we call this, it is always being called with with the billing project and then an immediate execution of one of the member functions or a reference to the member function. In other words, we never just store the result of the wrapper as `currentBillingProject` and invoke a call from there. And this is further complicated by the fact that we don't need a signal in one of the cases.

In this particular case, I am in favor of removing the wrapper. It does increase the arity of the functions, but it seems closer to what the function is doing and a bit simpler.

File: src/pages/billing/Project.js (hunk 9 above)
Comment (petesantos) [reply src/pages/billing/Project.js:335]: Looking through the code, while we do use `_.curry` we do not have a huge number of cases where we curry functions. Though we do take advantage of the auto currying in the `lodash/fp` variant, so it should become familiar to those using the codebase.

But it seems that most of the functions we curry in the codebase are wrappers (e.g. `withSomething`) or utilities intended to be used similar to an auto-curried lodash function. In this specific case, I like partial as it is clear that we are not fully executing a function - which is especially helpful in `addFunction` and `saveFunction`

Although this is tied to the concept of the `project` wrapper in the `Ajax` file.

Personally, I am ok with this. This would be another interesting topic for our next ui review discussion.

**Review by petesantos — APPROVED** [review]
Added the currying and ternary discussion to the next UI meeting
